### ISSUE TITLE
Perf bottleneck of `FIONREAD` OCALL during PAL attributes read

### ISSUE THREAD

**Issue by dimakuv** [gramineproject/gramine#893]
## Description of the problem

We have a perf bottleneck in Linux-SGX PAL. In particular, to get the attributes of the PAL handle (via `PalStreamAttributesQueryByHandle()`), we *always* perform one SGX OCALL to get the pending bytes-to-read. We have these four places in Linux-SGX code:
https://github.com/gramineproject/gramine/blob/f112c819ea030bbc144e712e17079d010a347444/pal/src/host/linux-sgx/pal_eventfd.c#L113
https://github.com/gramineproject/gramine/blob/48f4660a4b5e3e469922bb8cef7e99a336c06b59/pal/src/host/linux-sgx/pal_process.c#L330
https://github.com/gramineproject/gramine/blob/f112c819ea030bbc144e712e17079d010a347444/pal/src/host/linux-sgx/pal_pipes.c#L495
https://github.com/gramineproject/gramine/blob/48f4660a4b5e3e469922bb8cef7e99a336c06b59/pal/src/host/linux-sgx/pal_sockets.c#L284

I think we may ignore the first two places: they are about `eventfd` and `process` PAL objects. These objects are not that frequently used, so this shouldn't introduce visible perf overhead in real-world applications.

But the last two places are pretty serious: real-world applications may read `pipes` and `sockets` attributes quite frequently. Doing one EEXIT per each read-attributes operation can introduce a significant performance hit on e.g. web servers.

## Possible solution

We should probably introduce an additional argument to `PalStreamAttributesQueryByHandle()` as to which attributes we are interested in. Currently, Gramine just updates/reads all attributes at once, even if the code actually wants to set/get just one attribute.

**Comment by svenkata9** [comment]
[nginx_gsgx_strace.txt](https://github.com/gramineproject/gramine/files/9507172/nginx_gsgx_strace.txt)

(Context of this PR was poor perf with NGINX. Thanks @dimakuv for creating an issue).
Attaching the NGINX log that shows a huge hit in perf w.r.t. native because of these IOCTLs with FIONREAD.

**Comment by boryspoplawski** [comment]
Who analyzed this issue and how did they come to these conclusions?

My analysis of the provided log would be quite, quite different. There are 24 calls to `ioctl(FIONREAD)` which should take close to 0 time. The performance hit comes from `ioctl`s for creating the enclave (mostly adding pages).

**Comment by boryspoplawski** [comment]
I have a hard time believing Nginx does `setsockopt` so frequent that this is a perf bottleneck.

**Comment by dimakuv** [comment]
We should be talking about the usages of `PalStreamAttributesQueryByHandle()`.

**Comment by boryspoplawski** [comment]
Which are:
- `setsockopt`,
- actually querying pending size,
- changing flags of a fd,
- emulating `MSG_DONTWAIT` for UNIX domain sockets - newly added, Nginx didnt use it.

**Comment by dimakuv** [comment]
@boryspoplawski Yes, these are the usages. I think also `getsockopt`. Well, need to verify all the places that lead to calling the `attrquerybyhdl()` callback, I think there were a couple more (inside the PAL code itself).

Just to reiterate, I have a strong feeling that performing an FIONREAD OCALL is suboptimal.

**Comment by boryspoplawski** [comment]
Of course, it definitely is suboptimal. But I would think you cannot notice it on any real workload. Could you share any perf/strace/whatever showing otherwise?

**Comment by dimakuv** [comment]
Someone (maybe me) needs to create a reproducer and show perf numbers (+ show perf numbers on a real workload like Nginx).

This is low priority for now though.
